File: python/test/test_tiling_schema.py
# python/test/test_tiling_schema.py
"""
Unit tests for TilingSchema.

Run:
    cd /Volumes/GM9/code/Ascend-MLIR
    python3 -m pytest python/test/test_tiling_schema.py -v

The fixture is examples/broadcast-add-reduce/tiling_space.json.
It has 6 int64 fields: TB_M, TB_N, dim_arg0_0, dim_arg1_1, dim_arg0_1, dim_arg1_0.
"""
import json
import struct
import tempfile
from pathlib import Path

import pytest

import sys
sys.path.insert(0, str(Path(__file__).parent.parent.parent))
from python.runtime.tiling_schema import TilingSchema

FIXTURE = (
    Path(__file__).parent.parent.parent
    / "examples/broadcast-add-reduce/tiling_space.json"
)

# ── Helpers ────────────────────────────────────────────────────────────────────

def make_json(params: list[dict]) -> Path:
    """Write a temporary tiling_space.json and return its path."""
    tmp = tempfile.NamedTemporaryFile(
        suffix=".json", mode="w", delete=False
    )
    json.dump({"kernel": "test", "tiling_params": params}, tmp)
    tmp.close()
    return Path(tmp.name)


# ── from_json ─────────────────────────────────────────────────────────────────

def test_from_json_roundtrip():
    schema = TilingSchema.from_json(FIXTURE)
    assert schema.field_names == [
        "TB_M", "TB_N", "dim_arg0_0", "dim_arg1_1", "dim_arg0_1", "dim_arg1_0"
    ]
    assert len(schema) == 6
    assert all(f["type"] == "int64" for f in schema.fields)


def test_from_json_file_not_found():
    with pytest.raises(FileNotFoundError, match="not found"):
        TilingSchema.from_json("/nonexistent/path/schema.json")


def test_from_json_malformed_json():
    with tempfile.NamedTemporaryFile(suffix=".json", mode="w", delete=False) as f:
        f.write("{ not valid json")
        path = Path(f.name)
    with pytest.raises(ValueError, match="JSON parse error"):
        TilingSchema.from_json(path)


def test_from_json_missing_tiling_params():
    with tempfile.NamedTemporaryFile(suffix=".json", mode="w", delete=False) as f:
        json.dump({"kernel": "test"}, f)
        path = Path(f.name)
    with pytest.raises(ValueError, match="missing 'tiling_params'"):
        TilingSchema.from_json(path)


def test_from_json_type_defaults_to_int64():
    path = make_json([{"name": "X"}])  # no "type" key
    schema = TilingSchema.from_json(path)
    assert schema.fields[0]["type"] == "int64"


# ── pack ──────────────────────────────────────────────────────────────────────

def test_pack_happy_path_all_int64():
    schema = TilingSchema.from_json(FIXTURE)
    data = schema.pack(
        TB_M=16, TB_N=16,
        dim_arg0_0=64, dim_arg1_1=64,
        dim_arg0_1=64, dim_arg1_0=64,
    )
    # 6 × 8 bytes = 48
    assert len(data) == 48
    vals = struct.unpack_from("<6q", data)
    assert vals == (16, 16, 64, 64, 64, 64)


def test_pack_int32_vs_int64():
    path = make_json([
        {"name": "A", "type": "int64"},
        {"name": "B", "type": "int32"},
        {"name": "C", "type": "int64"},
    ])
    schema = TilingSchema.from_json(path)
    data = schema.pack(A=1, B=2, C=3)
    # int64(8) + int32(4) + int64(8) = 20 bytes
    assert len(data) == 20
    a = struct.unpack_from("<q", data, 0)[0]
    b = struct.unpack_from("<i", data, 8)[0]
    c = struct.unpack_from("<q", data, 12)[0]
    assert (a, b, c) == (1, 2, 3)


def test_pack_int32_alias():
    """int32_t is accepted as a type alias for int32."""
    path = make_json([{"name": "X", "type": "int32_t"}])
    schema = TilingSchema.from_json(path)
    data = schema.pack(X=42)
    assert len(data) == 4
    assert struct.unpack_from("<i", data)[0] == 42


def test_pack_missing_field():
    schema = TilingSchema.from_json(FIXTURE)
    with pytest.raises(ValueError) as exc:
        schema.pack(TB_M=16, TB_N=16)  # missing 4 fields
    msg = str(exc.value)
    assert "missing tiling fields" in msg
    assert "dim_arg0_0" in msg
    assert "expected" in msg


def test_pack_extra_field():
    schema = TilingSchema.from_json(FIXTURE)
    with pytest.raises(ValueError) as exc:
        schema.pack(
            TB_M=16, TB_N=16,
            dim_arg0_0=64, dim_arg1_1=64,
            dim_arg0_1=64, dim_arg1_0=64,
            EXTRA=99,
        )
    msg = str(exc.value)
    assert "unexpected tiling fields" in msg
    assert "EXTRA" in msg


def test_pack_field_order_in_output_matches_schema():
    """Output byte order follows schema declaration order, not kwargs order."""
    path = make_json([
        {"name": "FIRST", "type": "int64"},
        {"name": "SECOND", "type": "int64"},
    ])
    schema = TilingSchema.from_json(path)
    # Intentionally pass kwargs in reverse order
    data = schema.pack(SECOND=200, FIRST=100)
    first, second = struct.unpack_from("<2q", data)
    assert first == 100
    assert second == 200


def test_repr():
    schema = TilingSchema.from_json(FIXTURE)
    r = repr(schema)
    assert "TB_M" in r
    assert "TilingSchema" in r

File: test/tools/runtime/test_tiling_schema.cpp
// test/tools/runtime/test_tiling_schema.cpp
//
// Unit tests for TilingSchema (fromJson + pack).
// Does NOT require a simulator or .bin file.
//
// Build (on xvm):
//   LLVM_BUILD=~/code/llvm-project/llvm/build
//   cd /home/niu/code/Ascend-MLIR/build
//   cmake --build . --target AscendCRuntime -j4
//   cd ..
//   g++ -std=c++17 -I include/ -I $LLVM_BUILD/include \
//       -I ~/code/llvm-project/llvm/include \
//       test/tools/runtime/test_tiling_schema.cpp \
//       build/lib/libAscendCRuntime.a \
//       $($LLVM_BUILD/bin/llvm-config --ldflags --libs support) \
//       -ldl -o /tmp/test_tiling_schema
//   /tmp/test_tiling_schema <path/to/tiling_space.json>
//
// The fixture used is examples/broadcast-add-reduce/tiling_space.json.
// It defines 6 int64 fields: TB_M, TB_N, dim_arg0_0, dim_arg1_1, dim_arg0_1, dim_arg1_0.
//
#include "Runtime/TilingSchema.h"
#include "llvm/Support/raw_ostream.h"
#include <cassert>
#include <cstring>
#include <fstream>
#include <string>
#include <vector>

using namespace mlir::runtime;

static int g_pass = 0, g_fail = 0;

#define EXPECT(cond, msg)                                              \
  do {                                                                 \
    if (cond) {                                                        \
      llvm::outs() << "  PASS: " << (msg) << "\n";                    \
      ++g_pass;                                                        \
    } else {                                                           \
      llvm::errs() << "  FAIL: " << (msg) << "\n";                    \
      ++g_fail;                                                        \
    }                                                                  \
  } while (0)

static std::string writeMixedJson(const std::string& path) {
  std::ofstream f(path);
  f << R"({
  "kernel": "test_kernel",
  "tiling_params": [
    {"name": "A", "type": "int64"},
    {"name": "B", "type": "int32"},
    {"name": "C", "type": "int64"}
  ]
})";
  return path;
}

static std::string writeMalformedJson(const std::string& path) {
  std::ofstream f(path);
  f << "{ not valid json";
  return path;
}

static void testFromJsonRoundtrip(const std::string& fixture) {
  llvm::outs() << "\n[fromJson roundtrip]\n";
  auto s = TilingSchema::fromJson(fixture);
  EXPECT(!s.takeError(), "fromJson succeeds on valid fixture");

  auto s2 = TilingSchema::fromJson(fixture);
  if (!s2) { llvm::consumeError(s2.takeError()); return; }

  EXPECT(s2->size() == 6, "fixture has 6 fields");
  EXPECT(s2->fields()[0].name == "TB_M",       "fields[0].name == TB_M");
  EXPECT(s2->fields()[0].type == "int64",      "fields[0].type == int64");
  EXPECT(s2->fields()[1].name == "TB_N",       "fields[1].name == TB_N");
  EXPECT(s2->fields()[2].name == "dim_arg0_0", "fields[2].name == dim_arg0_0");
  EXPECT(s2->fields()[5].name == "dim_arg1_0", "fields[5].name == dim_arg1_0");
}

static void testPackHappyPath(const std::string& fixture) {
  llvm::outs() << "\n[pack happy path]\n";
  auto s = TilingSchema::fromJson(fixture);
  if (!s) { llvm::consumeError(s.takeError()); return; }

  using P = std::pair<std::string, int64_t>;
  auto result = s->pack({
    P{"TB_M",       16},
    P{"TB_N",       16},
    P{"dim_arg0_0", 64},
    P{"dim_arg1_1", 64},
    P{"dim_arg0_1", 64},
    P{"dim_arg1_0", 64},
  });
  EXPECT(!result.takeError(), "pack succeeds with correct params");

  auto result2 = TilingSchema::fromJson(fixture);
  if (!result2) { llvm::consumeError(result2.takeError()); return; }
  auto bytes = result2->pack({
    P{"TB_M",       16},
    P{"TB_N",       16},
    P{"dim_arg0_0", 64},
    P{"dim_arg1_1", 64},
    P{"dim_arg0_1", 64},
    P{"dim_arg1_0", 64},
  });
  if (!bytes) { llvm::consumeError(bytes.takeError()); return; }

  EXPECT(bytes->size() == 48, "6 int64 fields -> 48 bytes");

  int64_t tb_m = 0;
  std::memcpy(&tb_m, bytes->data(), 8);
  EXPECT(tb_m == 16, "bytes[0..7] == 16 (TB_M)");

  int64_t last = 0;
  std::memcpy(&last, bytes->data() + 40, 8);
  EXPECT(last == 64, "bytes[40..47] == 64 (dim_arg1_0)");
}

static void testPackWrongCount(const std::string& fixture) {
  llvm::outs() << "\n[pack wrong count]\n";
  auto s = TilingSchema::fromJson(fixture);
  if (!s) { llvm::consumeError(s.takeError()); return; }

  using P = std::pair<std::string, int64_t>;
  auto result = s->pack({P{"TB_M", 16}, P{"TB_N", 16}});
  auto err = result.takeError();
  EXPECT((bool)err, "pack returns error when count mismatches");
  std::string msg = llvm::toString(std::move(err));
  EXPECT(msg.find("count mismatch") != std::string::npos,
         "error message contains 'count mismatch'");
}

static void testPackWrongName(const std::string& fixture) {
  llvm::outs() << "\n[pack wrong name]\n";
  auto s = TilingSchema::fromJson(fixture);
  if (!s) { llvm::consumeError(s.takeError()); return; }

  using P = std::pair<std::string, int64_t>;
  auto result = s->pack({
    P{"WRONG",      16},
    P{"TB_N",       16},
    P{"dim_arg0_0", 64},
    P{"dim_arg1_1", 64},
    P{"dim_arg0_1", 64},
    P{"dim_arg1_0", 64},
  });
  auto err = result.takeError();
  EXPECT((bool)err, "pack returns error when name mismatches");
  std::string msg = llvm::toString(std::move(err));
  EXPECT(msg.find("name mismatch") != std::string::npos,
         "error message contains 'name mismatch'");
  EXPECT(msg.find("WRONG") != std::string::npos,
         "error message contains the bad name 'WRONG'");
  EXPECT(msg.find("TB_M") != std::string::npos,
         "error message contains the expected name 'TB_M'");
}

static void testFromJsonFileNotFound() {
  llvm::outs() << "\n[fromJson file not found]\n";
  auto s = TilingSchema::fromJson("/nonexistent/path/schema.json");
  auto err = s.takeError();
  EXPECT((bool)err, "fromJson returns error for missing file");
  llvm::consumeError(std::move(err));
}

static void testFromJsonMalformed() {
  llvm::outs() << "\n[fromJson malformed JSON]\n";
  std::string path = "/tmp/malformed_tiling.json";
  writeMalformedJson(path);
  auto s = TilingSchema::fromJson(path);
  auto err = s.takeError();
  EXPECT((bool)err, "fromJson returns error for malformed JSON");
  llvm::consumeError(std::move(err));
}

static void testPackMixedTypes() {
  llvm::outs() << "\n[pack mixed int32/int64 types]\n";
  std::string path = "/tmp/mixed_tiling.json";
  writeMixedJson(path);
  auto s = TilingSchema::fromJson(path);
  if (!s) { llvm::consumeError(s.takeError()); return; }
  EXPECT(s->size() == 3, "mixed fixture has 3 fields");
  EXPECT(s->fields()[0].type == "int64", "A is int64");
  EXPECT(s->fields()[1].type == "int32", "B is int32");
  EXPECT(s->fields()[2].type == "int64", "C is int64");

  using P = std::pair<std::string, int64_t>;
  auto bytes = s->pack({P{"A", 1}, P{"B", 2}, P{"C", 3}});
  if (!bytes) { llvm::consumeError(bytes.takeError()); return; }

  EXPECT(bytes->size() == 20, "int64+int32+int64 = 20 bytes");

  int64_t a = 0; std::memcpy(&a, bytes->data() + 0, 8);
  int32_t b = 0; std::memcpy(&b, bytes->data() + 8, 4);
  int64_t c = 0; std::memcpy(&c, bytes->data() + 12, 8);
  EXPECT(a == 1, "A=1");
  EXPECT(b == 2, "B=2");
  EXPECT(c == 3, "C=3");
}

int main(int argc, char** argv) {
  if (argc < 2) {
    llvm::errs() << "Usage: " << argv[0]
                 << " <path/to/tiling_space.json>\n";
    return 2;
  }
  std::string fixture = argv[1];

  testFromJsonRoundtrip(fixture);
  testPackHappyPath(fixture);
  testPackWrongCount(fixture);
  testPackWrongName(fixture);
  testFromJsonFileNotFound();
  testFromJsonMalformed();
  testPackMixedTypes();

  llvm::outs() << "\n" << g_pass << " passed, " << g_fail << " failed\n";
  return g_fail > 0 ? 1 : 0;
}
